Project scoping fixes - no recompilation of contract code on tests

pdobacz opened the pull request:
### What was wrong?

see https://github.com/pipermerriam/populus/issues/184

### How was it fixed?

`scope='session'` for `project` fixture and using pytest `request.config.cache` to store the `Project`'s compiled contracts across test runs

#### Cute Animal Picture

![](http://justcuteanimals.com/wp-content/uploads/2016/09/highland-cows-baby-animal-pictures-pics.jpg)

pdobacz commented:
<sigh> I've totally missed the 1.0.0 and just now seen that not only is the `session` part done, but also the cross-testrun part won't run anymore, as `Project.compiled_contracts` switched from hashes to `os.stat`, that pytest's cache refuses to store

Hmmm, however, I'd be glad to have something like this (i.e. caching compiled contracts across test runs). Any thoughts @pipermerriam?

pipermerriam commented:
@pdobacz sorry for any confusion.  Some of these changes were quickly tossed in to remove some significant slowdowns in compilation and I fixed the session scoping in the process.

I'm going to be traveling over the next day or two so apologies if I'm slow to respond.  I'm supportive of any test run speed improvements you're able to provide.

pdobacz commented:
@pipermerriam np. I have updated the PR, so that it should make more sense now

pipermerriam commented on populus/plugin.py:
```diff
@@ -5,12 +5,23 @@
+    # use pytest cache to preset the sessions project to recently compiled contracts
+    contracts = request.config.cache.get(CACHE_KEY_CONTRACTS, None)
+    mtime = request.config.cache.get(CACHE_KEY_MTIME, None)
+    project = Project(cached_contracts=contracts, contracts_mtime=mtime)
```
Rather than complicating the constructor of the `Project` to work with this fixture what do you think about just manually setting these values on the `project` instance once instantiated?  I think I'm comfortable with that behavior as long as there is a basic test to cover that it works as expected.

pipermerriam commented on populus/project.py:
```diff
@@ -140,12 +149,21 @@ def get_source_file_hash(self):
 
-        if self._cached_compiled_contracts_stat != source_stat:
-            self._cached_compiled_contracts_stat = source_stat
+        if no_cached or self._cached_compiled_contracts_mtime < source_mtime:
```
Maybe wrap this up into it's own helper method?  Something like `self.contract_sources_have_changed()` or whatever name you think is appropriate.

pdobacz commented:
Thank you for the tips @pipermerriam. I updated the PR, the build fails now, AFAIK, due to `ethereum-testerc-client` hitting 1.2.2, but I'm not sure 100%. Most likely the fail is unrelated to the PR itself.

I am setting the `Project`'s cache using a function and I've added a test for its behavior. I wanted to avoid touching the `_` variables in `Project` directly, I hope that's what you meant.

pipermerriam commented:
Thanks for this.  Going to merge via https://github.com/pipermerriam/populus/pull/188
